graylog2-server/src/main/java/org/graylog2/database/suggestions/MongoEntitySuggestionService.java

```diff
@@ -56,42 +57,60 @@ public EntitySuggestionResponse suggest(final String collection,
                                             final int page,
                                             final int perPage,
                                             final Subject subject) {
-
-        if (!hasAllPermission(subject) && !hasReadPermissionForWholeCollection(subject, collection)) {
-            return new EntitySuggestionResponse(List.of(),
-                    PaginatedList.PaginationInfo.create(0, 0, page, perPage));
-        }
-
         final MongoCollection<Document> mongoCollection = mongoConnection.getMongoDatabase().getCollection(collection);
 
-        Bson bsonFilter = (query != null && !query.isEmpty()) ?
-                Filters.regex(valueColumn, query, "i") :
-                new BsonDocument();
+        final var bsonFilter = !Strings.isNullOrEmpty(query)
+                ? Filters.regex(valueColumn, query, "i")
+                : Filters.empty();
 
-        final FindIterable<Document> documents = mongoCollection
+        final var resultWithoutPagination = mongoCollection
                 .find(bsonFilter)
                 .projection(Projections.include(valueColumn))
-                .sort(Sorts.ascending(valueColumn))
-                .limit(perPage)
-                .skip((page - 1) * perPage);
+                .sort(Sorts.ascending(valueColumn));
+
+        final var userCanReadAllEntities = hasAllPermission(subject) || hasReadPermissionForWholeCollection(subject, collection);
+        final var skip = (page - 1) * perPage;
+        final var checkPermission = createPermissionCheck(subject, collection);
+        final var documents = userCanReadAllEntities
+                ? mongoPaginate(resultWithoutPagination, perPage, skip)
+                : paginateWithPermissionCheck(resultWithoutPagination, perPage, skip, checkPermission);
 
         final List<EntitySuggestion> suggestions = documents
                 .map(doc ->
                         new EntitySuggestion(
-                                doc.getObjectId("_id").toString(),
+                                doc.getObjectId(ID_FIELD).toString(),
                                 doc.getString(valueColumn)
                         )
                 )
-                .into(new ArrayList<>());
+                .toList();
 
-        final long total = mongoCollection.countDocuments(bsonFilter);
+        final long total = userCanReadAllEntities
+                ? mongoCollection.countDocuments(bsonFilter)
+                : MongoUtils.stream(mongoCollection.find(bsonFilter)).filter(checkPermission).count();
 
```

Comment:
I am not sure if projection won't be a good idea here as well. Entities have many fields, we care only about id in permission check, don't we?

Reply:
You are right!